File: src/main/java/me/tomasan7/jecnadesktop/ui/component/LessonView.java
package me.tomasan7.jecnadesktop.ui.component;

import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.StackPane;
import me.tomasan7.jecnadesktop.data.Lesson;

public class LessonView extends AnchorPane
{
	/** The distance of the corner {@link Label labels} from the sides. */
	private static final double CORNER_DISTANCE = 5d;

	private final Lesson lesson;

	private final Label subjectLabel;
	private final StackPane subjectLabelContainer;
	private final Label teacherLabel;
	private final Label classroomLabel;
	private final Label groupLabel;

	public LessonView (Lesson lesson)
	{
		this.lesson = lesson;
		this.subjectLabel = new Label(lesson.subjectShort());
		this.subjectLabelContainer = new StackPane();
		this.teacherLabel = new Label(lesson.teacherShort());
		this.classroomLabel = new Label(lesson.classroom());
		this.groupLabel = lesson.group() != 0
				? new Label(String.valueOf(lesson.group()))
				: null;

		initSubjectLabel();
		initTeacherLabel();
		initClassroomLabel();
		initGroupLabel();

		getChildren().addAll(subjectLabelContainer, teacherLabel, classroomLabel);
		if (groupLabel != null)
			getChildren().add(groupLabel);

		getStyleClass().add("lesson-view");
		getStylesheets().add("/ui/component/LessonView.css");
	}

	private void initSubjectLabel ()
	{
		subjectLabel.getStyleClass().add("subject-label");

		fit(subjectLabelContainer);
		subjectLabelContainer.getChildren().add(subjectLabel);
		StackPane.setAlignment(subjectLabel, Pos.CENTER);
	}

	private void initTeacherLabel ()
	{
		teacherLabel.getStyleClass().add("teacher-label");

		AnchorPane.setLeftAnchor(teacherLabel, CORNER_DISTANCE);
		AnchorPane.setTopAnchor(teacherLabel, CORNER_DISTANCE);
	}

	private void initClassroomLabel ()
	{
		classroomLabel.getStyleClass().add("classroom-label");

		AnchorPane.setRightAnchor(classroomLabel, CORNER_DISTANCE);
		AnchorPane.setTopAnchor(classroomLabel, CORNER_DISTANCE);
	}

	private void initGroupLabel ()
	{
		if (groupLabel == null)
			return;

		groupLabel.getStyleClass().add("group-label");

		AnchorPane.setRightAnchor(groupLabel, CORNER_DISTANCE);
		AnchorPane.setBottomAnchor(groupLabel, CORNER_DISTANCE);
	}

	/**
	 * Fits the {@link Node} so it takes 100% of this {@link AnchorPane}.
	 * @param node The {@link Node} to fit.
	 */
	private void fit (Node node)
	{
		AnchorPane.setTopAnchor(node, 0d);
		AnchorPane.setRightAnchor(node, 0d);
		AnchorPane.setBottomAnchor(node, 0d);
		AnchorPane.setLeftAnchor(node, 0d);
	}

	public Lesson getLesson ()
	{
		return lesson;
	}
}

File: src/main/kotlin/me/tomasan7/jecnadesktop/ui/component/LessonView.kt
package me.tomasan7.jecnadesktop.ui.component

import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.control.Label
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.StackPane
import me.tomasan7.jecnaapi.data.Lesson

class LessonView(val lesson: Lesson) : AnchorPane()
{

    init
    {
        children.addAll(createSubjectLabel(),
                        createTeacherLabel(),
                        createClassroomLabel())

        if (lesson.group != 0)
            children.add(createGroupLabel())

        styleClass.add("lesson-view")
        stylesheets.add("/ui/component/LessonView.css")
    }

    private fun createSubjectLabel(): StackPane
    {
        val subjectLabel = Label(lesson.subjectShort)
        subjectLabel.styleClass.add("subject-label")

        val subjectLabelContainer = StackPane()

        fit(subjectLabelContainer)
        subjectLabelContainer.children.add(subjectLabel)
        StackPane.setAlignment(subjectLabel, Pos.CENTER)

        return subjectLabelContainer
    }

    private fun createTeacherLabel(): Label
    {
        val teacherLabel = Label(lesson.teacherShort)
        teacherLabel.styleClass.add("teacher-label")

        setLeftAnchor(teacherLabel, CORNER_DISTANCE)
        setTopAnchor(teacherLabel, CORNER_DISTANCE)

        return teacherLabel
    }

    private fun createClassroomLabel(): Label
    {
        val classroomLabel = Label(lesson.classroom)
        classroomLabel.styleClass.add("classroom-label")

        setRightAnchor(classroomLabel, CORNER_DISTANCE)
        setTopAnchor(classroomLabel, CORNER_DISTANCE)

        return classroomLabel
    }

    private fun createGroupLabel(): Label
    {
        val groupLabel = Label(lesson.group.toString())
        groupLabel.styleClass.add("group-label")

        setRightAnchor(groupLabel, CORNER_DISTANCE)
        setBottomAnchor(groupLabel, CORNER_DISTANCE)

        return groupLabel
    }

    /**
     * Fits the [Node] so it takes 100% of this [AnchorPane].
     *
     * @param node The [Node] to fit.
     */
    private fun fit(node: Node)
    {
        setTopAnchor(node, 0.0)
        setRightAnchor(node, 0.0)
        setBottomAnchor(node, 0.0)
        setLeftAnchor(node, 0.0)
    }

    companion object
    {
        /** The distance of the corner [labels][Label] from the sides.  */
        private const val CORNER_DISTANCE = 5.0
    }
}